src/util/futurepool.rs

```diff
@@ -0,0 +1,227 @@
+use std::thread;
+use std::time;
+use futures::Future;
+use futures_cpupool as cpupool;
+
+use util;
+use util::time::Instant;
+use util::collections::HashMap;
+
+pub trait Context: fmt::Debug + Send + 'static {
+    fn on_tick(&mut self) {}
+}
+
+#[derive(Debug)]
+/// A wrapper to wrap context to provide `on_tick` feature.
+struct ContextOuter<T: Context> {
+    tick_interval: time::Duration,
+    /// TODO: Can be replace by UnsafeCell to eliminate the runtime cost of borrow checking.
+    inner: cell::RefCell<T>,
+    last_tick: cell::Cell<Option<Instant>>,
+}
+
+impl<T: Context> ContextOuter<T> {
+    fn new(context: T, tick_interval: time::Duration) -> ContextOuter<T> {
+        ContextOuter {
+            tick_interval,
+            inner: cell::RefCell::new(context),
+            last_tick: cell::Cell::new(None),
+        }
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
+            // set last_tick when the first future is resolved
+            self.last_tick.set(Some(now));
+            return;
+        }
+        if now.duration_since(last_tick.unwrap()) < self.tick_interval {
+            return;
+        }
+        self.last_tick.set(Some(now));
+        self.inner.borrow_mut().on_tick();
+    }
+}
+
+/// Each `ContextOuter` instance is invoked individually for each thread
+/// so that we mark it as Sync.
+unsafe impl<T: Context> Sync for ContextOuter<T> {}
+
+#[derive(Debug)]
+/// A future thread pool that supports `on_tick` for each thread.
+pub struct FuturePool<T: Context> {
+    pool: cpupool::CpuPool,
+    contexts: sync::Arc<HashMap<thread::ThreadId, ContextOuter<T>>>,
+}
+
+impl<T: Context> Clone for FuturePool<T> {
+    fn clone(&self) -> FuturePool<T> {
+        FuturePool::<T> {
+            pool: self.pool.clone(),
+            contexts: sync::Arc::clone(&self.contexts),
+        }
+    }
+}
+
+impl<T: Context> util::AssertSend for FuturePool<T> {}
+impl<T: Context> util::AssertSync for FuturePool<T> {}
+
+impl<T: Context> FuturePool<T> {
+    pub fn new<F>(
+        pool_size: usize,
+        stack_size: usize,
+        name_prefix: &str,
+        tick_interval: time::Duration,
+        context_factory: F,
+    ) -> FuturePool<T>
+    where
+        F: Send + 'static + Fn(thread::ThreadId) -> T,
+    {
+        let (tx, rx) = mpsc::sync_channel(pool_size);
+        let pool = cpupool::Builder::new()
+            .pool_size(pool_size)
+            .stack_size(stack_size)
+            .name_prefix(name_prefix)
+            .after_start(move || {
+                // We only need to know each thread's id and we can build context later
+                // by invoking `context_factory` in a !Sync way.
+                let thread_id = thread::current().id();
+                tx.send(thread_id).unwrap();
+            })
+            .create();
+        let contexts = rx.into_iter()
+            .map(|thread_id| {
```

Comment:
Here please make it more clear that must receive `poolsize` messages.

Reply:
done